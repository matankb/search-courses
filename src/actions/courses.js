import axios from "axios";

export const setCourses = (courses) => ({
    type: "SET_COURSES",
    courses
});

export const startSetCourses = () => {
    return (dispatch, getState) => {
        return axios.get("/courseslist")
            .then((response) => {
                dispatch(setCourses(response.data));
            })
            .catch((e) => {
                console.log(e);
            });
    };
};