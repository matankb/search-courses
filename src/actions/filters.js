export const setSearchQuery = (searchQuery = "") => ({
    type: "SET_SEARCH_QUERY",
    searchQuery
});

export const setCurrentPage = (currentPage = 1) => ({
    type: "SET_CURRENT_PAGE",
    currentPage
});