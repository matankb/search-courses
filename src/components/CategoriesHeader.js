import React from "react";

export const CategoriesHeader = (props) => {
    return (
        <div className="search_results_page__categories">
            <div className="search-results-page__top__title">{props.title}: <span>{props.category}</span></div>
        </div>
    );
};

export default CategoriesHeader;