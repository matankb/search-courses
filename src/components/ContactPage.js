import React from "react";
import axios from "axios";
import MetaTags from 'react-meta-tags';


export class ContactPage extends React.Component {
    state = {
        name: "",
        email: "",
        text: "",
        isFormValidated: false,
        error: false,
        formSendInProgress: false,
        isMessageSent: false
    }

    onNameChange (e) {
        const name = e.target.value;
        this.setState(() => ({name}));
    }

    onEmailChange (e) {
        const email = e.target.value;
        this.setState(() => ({email}));
    }

    onTextChange (e) {
        const text = e.target.value;
        this.setState(() => ({text}));
    }

    isFormValidated () {
        const isNameFieldNotEmpty = this.state.name.trim() !== "";
        const isEmailValidated = /^.+@.+\..+$/.test(this.state.email);
        const isTextFieldNotEmpty = this.state.text.trim() !== "";

        const isFormValidated = isEmailValidated && isNameFieldNotEmpty && isTextFieldNotEmpty;
        this.setState(() => ({isFormValidated}));
    }

    handleOnSubmit (e) {
        e.preventDefault();
        this.setState(() => ({formSendInProgress: true}));

        axios.post("/api/contactus", {
            name: this.state.name,
            email: this.state.email,
            text: this.state.text
        }).then((res) => {
            this.setState(() => ({formSendInProgress: false}));
            this.setState(() => ({isMessageSent: true}));
        }).catch((err) => {
            this.setState(() => ({error: true}));
            this.setState(() => ({formSendInProgress: false}));
            console.log("Error sending email");
        });
    }

    render () {
        return (
            <div className="contact-form-container">
                <MetaTags>
                    <title>Contact Us | CourseSearch</title>
                    <meta name="description" content="Do you have any questions? Get in contact with us." />
                    <meta property="og:title" content="Contact Us | CourseSearch" />
                    <meta property="og:url" content="https://www.coursesearch.net/contact"/>
                    <meta name="twitter:title" content="Contact Us | CourseSearch"/>
                </MetaTags>
                
                <form
                    className="contact-form"
                    onKeyUp={this.isFormValidated.bind(this)}
                    onSubmit={this.handleOnSubmit.bind(this)}
                    >
                        <h2>Contact form</h2>
                        {this.state.error && <p className="error-message">Error while sending your message</p>}
                        <input
                            type="text"
                            value={this.state.name}
                            onChange={this.onNameChange.bind(this)}
                            placeholder="Your name *"
                            className="text-input contact-form__input"
                        />
                        <input
                            type="text"
                            value={this.state.email}
                            onChange={this.onEmailChange.bind(this)}
                            placeholder="Your email adress *"
                            className="text-input contact-form__input"
                        />
                        <textarea
                            type="textarea"
                            value={this.state.text}
                            onChange={this.onTextChange.bind(this)}
                            placeholder="Type your message here... *"
                            className="textarea contact-form__input"
                        />
                        <div>
                            <button className="button" type="submit" disabled={!this.state.isFormValidated || this.state.formSendInProgress || this.state.isMessageSent} >
                                {this.state.formSendInProgress ? "Sending..." : "Send"} 
                            </button>
                        </div>
                        {this.state.isMessageSent && <p className="success-message">Your message has been sent! We will get back to you as soon as possible</p>}
                </form>
            </div>
        );
    }
};

export default ContactPage;