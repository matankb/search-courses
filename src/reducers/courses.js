const coursesReducerDefaultState = {
    courseList: [],
    hasFoundResults: true
};

const coursesReducer = (state = coursesReducerDefaultState, action) => {
    switch(action.type) {
        case "SET_COURSES":
            return {
                ...state,
                courseList: action.courseList,
            };
        case "SET_HAS_FOUND_RESULTS":
            return {
                ...state,
                hasFoundResults: action.hasFoundResults
            }
        default: 
            return state;
    }
};

export default coursesReducer;