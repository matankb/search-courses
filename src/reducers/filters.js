const filtersDefaultState = {
    searchQuery: "",
    currentPage: 1,
    currentCategory: "",
    sortBy: "relevance"
};

const filtersReducer = (state = filtersDefaultState, action) => {
    switch(action.type) {
        case "SET_SEARCH_QUERY":
            return {
                ...state,
                searchQuery: action.searchQuery
            };
        case "SET_CURRENT_PAGE":
            return {
                ...state,
                currentPage: action.currentPage
            };
        case "SET_CURRENT_CATEGORY":
            return {
                ...state,
                currentCategory: action.currentCategory
            };
        case "SORT_BY_RELEVANCE":
            return {
                ...state,
                sortBy: "relevance"
            }
        case "SORT_BY_CHEAPEST":
            return {
                ...state,
                sortBy: "cheapest_first"
            }
        case "SORT_BY_EXPENSIVE":
            return {
                ...state,
                sortBy: "expensive_first"
            }
        default: 
            return state;
    }
};

export default filtersReducer;