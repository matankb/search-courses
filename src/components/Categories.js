import React from "react";
import { connect } from "react-redux";
import { startSetCourses } from "../actions/courses";
import { topics } from "../utils/categories";

export class Categories extends React.Component {
    constructor (props) {
        super(props);
        this.state = {
            topics : topics
        }
    }

    handleOnClick = (e) => {
        const item = this.state.topics.find((topic)=> {
            return e.target.dataset.id === topic.id;
        });

        this.props.startSetCourses(item.tags);
        this.props.history.push(`/search/${item.purpose}`);
    }

    render () {
        return (
            <div className="dropdown">
               <div>Categories</div>
               <div className="dropdown__menu">
                    <ul onClick={this.handleOnClick}>
                        {this.state.topics.map((topic) => {
                            return <li key={topic.id} data-id={topic.id}>{topic.name}</li>
                        })}
                    </ul>
                </div>
            </div>
        );
    }
};

const mapDispatchToProps = (dispatch) => ({
    startSetCourses: (textFilter) => dispatch(startSetCourses(textFilter))
});

export default connect(undefined, mapDispatchToProps)(Categories);

