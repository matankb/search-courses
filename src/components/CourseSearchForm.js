import React from "react";
import { connect } from "react-redux";
import { startSetCourses } from "../actions/courses";
import { setSearchQuery, setCurrentPage } from "../actions/filters";

export class CourseSearchForm extends React.Component {
    onTextChange = (e) => {
        const textFilter = e.target.value;

        this.props.setSearchQuery(textFilter);
    };
    handleOnSubmit = (e) => {
        e.preventDefault();
        const textFilter = this.props.searchQuery;

        if (textFilter.trim()) {
            this.props.onSubmit(textFilter);
            this.props.setSearchQuery(textFilter);
            this.props.setCurrentPage();
        }
    };
    render () {
        return (
            <form onSubmit={this.handleOnSubmit}>
                <input
                    value={this.props.searchQuery}
                    type="text"
                    autoFocus
                    placeholder="What are you looking for?"
                    onChange={this.onTextChange}
                >
                </input>
                <button type="submit">Search</button>
            </form>
        );
    }
}

const mapStateToProps = (state) => ({
    searchQuery: state.filters.searchQuery
});

const mapDispatchToProps = (dispatch) => ({
    startSetCourses: (textFilter) => dispatch(startSetCourses(textFilter)),
    setSearchQuery: (textFilter) => dispatch(setSearchQuery(textFilter)),
    setCurrentPage: (page) => dispatch(setCurrentPage(page))
});

export default connect(mapStateToProps, mapDispatchToProps)(CourseSearchForm);