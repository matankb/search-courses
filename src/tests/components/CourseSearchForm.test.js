import React from "react";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { CourseSearchForm } from "../../components/CourseSearchForm";
import { shallow } from "enzyme";
import { setSearchQuery } from "../../actions/filters";
import courses from "../fixtures/courses";

const uid = "this_is_my_test_uid";
const createMockStore = configureMockStore([thunk]);
let wrapper, onSubmit, store, searchQuery, setCurrentPage, setCurrentCategory;

beforeEach(() => {
    searchQuery = "Hello world";
    store = createMockStore();
    onSubmit = jest.fn();
    setCurrentPage=jest.fn();
    setCurrentCategory=jest.fn();
    wrapper = shallow(
        <CourseSearchForm
            searchQuery={searchQuery}
            onSubmit={onSubmit}
            setSearchQuery={setSearchQuery}
            setCurrentPage={setCurrentPage}
            setCurrentCategory={setCurrentCategory}
        />);
});

test("should render CourseSearchForm correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("should call onTextChange with correct data", async () => {
    const text = "Hello world!";

    await store.dispatch(setSearchQuery(text));

    expect(store.getActions()[0]).toEqual({
        type: "SET_SEARCH_QUERY",
        searchQuery: text
    });
});

test("should handle on text change", () => {
    const value = "Hakuna Matata!";

    wrapper.setState({textFilter: value});
    wrapper.find("input").simulate("change", {
        target: { value }
    });
    expect(wrapper.state("textFilter")).toBe(value);
});


test("should handle on submit correctly", async () => {
    const text = "Hello mama!";

    await store.dispatch(setSearchQuery(text));

    expect(store.getActions()[0]).toEqual({
        type: "SET_SEARCH_QUERY",
        searchQuery: text
    });

    wrapper.find("form").simulate("submit", {
        preventDefault: () => {}
    });
    expect(onSubmit).toHaveBeenCalledWith(searchQuery);
});