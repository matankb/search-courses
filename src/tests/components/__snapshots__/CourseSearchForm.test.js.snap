// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CourseSearchForm correctly 1`] = `
<form
  className="search-wrapper__form undefined"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input text-input--search"
    onChange={[Function]}
    placeholder="What are you looking for?"
    type="text"
    value="Hello world"
  />
  <button
    className="button button--search"
    type="submit"
  >
    <img
      alt="Magnifier glass icon"
      className="icon"
      src="../../images/search-icon.svg"
    />
  </button>
</form>
`;
